openc3/bin/openc3cli

```diff
@@ -648,6 +656,165 @@ def run_bridge(filename, params)
   end
 end
 
+def script_monitor(script_id)
+  ret_code = ERROR_CODE
+  OpenC3::RunningScriptWebSocketApi.new(id: script_id) do |api|
+    while (resp = api.read) do
+      # see ScriptRunner.vue for types and states
+      case resp['type']
+      when 'running', 'breakpoint', 'waiting'
+        puts resp.pretty_inspect
+      when 'error', 'fatal'
+        $script_left_waiting = ''
+        puts 'script failed'
+        break
+      when 'file'
+        puts "Filename #{resp['filename']} scope #{resp['scope']}"
+      when 'line'
+        fn = resp['filename'].nil? ? '<no file>' : resp['filename']
+        puts "At [#{fn}:#{resp['line_no']}] state [#{resp['state']}]"
+        if resp['state'] == 'error'
+          $script_left_waiting = ''
+          puts 'script failed'
+          break
+        end
+      when 'output'
+        puts resp['line']
+      when 'paused'
+        if resp['state'] == 'fatal'
+          $script_left_waiting = ''
+          puts 'script failed'
+          break
+        end
+      when 'time'
+        puts resp.pretty_inspect
+      when 'complete'
+        $script_left_waiting = ''
+        puts 'script complete'
+        ret_code = 0
+        break
+      else
+        puts resp.pretty_inspect
+      end
+    end
+  end
+  return ret_code
+end
+
+def cli_script_list(args=[])
+  path = ''
+  if (args[0] && args[0][0] == '/')
+    path = (args.shift)[1..-1]+'/'
+  end
+  scope = args[0]
+  scope ||= 'DEFAULT'
+  script_list(scope: scope).each do |script_name|
+    puts (script_name) if script_name.start_with?(path)
+  end
+  return 0
+end
+
+def cli_script_run(disconnect=false, environment={}, args=[])
+  #  we are limiting the wait for status, not the script run time
+  #  we make 0 mean 'forever'
+  ret_code = ERROR_CODE
+  wait_limit = 0
+  if (i = args.index('--wait'))
+    begin
+      args.delete('--wait')
+      # pull out the flag
+      seconds = args[i]
+      wait_limit = Integer(seconds, 10) # only decimal, ignore leading 0
+      args.delete_at(i)
+      # and its value
+    rescue ArgumentError
+      abort("  --wait requires a number of seconds to wait, not [#{seconds}]")
+    end
+  end
+  abort("No script file provided") if args[0].nil?
+  scope = args[1]
+  scope ||= 'DEFAULT'
+  id = script_run(args[0], disconnect: disconnect, environment: environment, scope: scope) # could raise
+  $script_left_waiting = "  Script #{args[0]} still running remotely.\n" # for Ctrl-C
+  if (wait_limit < 1) then
+    ret_code = script_monitor(id)
+  else
+    Timeout::timeout(wait_limit, nil, "--wait #{wait_limit} exceeded") do
```

Comment:
As a general rule we never use Timeout.   It ends up doing a Thread.kill which can cause all kinds of bad state leftover.    For the cli where it is about to die anyways, maybe it is ok, but setting a variable to tell your thread to stop by itself is way better.

Reply:
In general, I agree.  I was even more worried about blocking on the socket read, and didn't see any explicit thread handling nearby.  Of course, it's the ones you don't see that get you...